design/external-compaction.md

```diff
@@ -0,0 +1,249 @@
+This information could potentially be kept in memory in a structure of the form Map<QueueName, SortedMap<Priority, LinkedHashSet<Tserver>>> and when the coordinator starts it polls tservers and constructs the global queue summary map.  Periodically the coordinator could ask tservers for updated summaries and/or the reserveJob RPC call to tservers could return updated summary information.
+
+
+## Workflow
+
+When the coordinator polls a tserver to get its latest external queue summaries the following could happen:
+
+  1. The tserver returns the latest summary information for all of its external compaction queues.
+  2. The coordinator will add the tserver to priorities that it is not present in in the global queue summary map.
+  3. The coordinator will remove the tserver from priorities that is in in the global queue summary map, but it is not in the summaries just returned by the tserver
+  4. If needed the next tserver pointer in the global queue summary map will be updated.
+
+On request for work from a compactor
+  1. For the queue requested, the coordinator gets the next tserver from the global queue summary map.
+  2. The Coordinator contacts the tablet server to reserve job
+  3. Return compaction job information to compactor
+  4. This must handle concurrent requests properly while minimizing the amount of locking.
+
+## Metrics
+
+TServers will each maintain their own internal queues for external compaction queues with all tablets.  If each tserver emits metrics about their local queues, then sum of all tserver metrics will give information about the external queues across the cluster.
+
+Compactor should emit metrics about the number of compactions running and which queue they are running on.
+
+## Scenarios
+
+Scenario 1
+
+ 1. Tablet T1 has files F1,F4,F5,F6
+ 1. The compaction of files F4,F5,F6 is queued on external executor EE2 for Tablet T1
+ 1. A new file F7 arrives for T1
+ 1. The compaction of F4,F5,F6 on EE2 is canceled and new a compaction of F4,F5,F6,F7 is queued on EE2 for Tablet T1
+ 1. A new file F8 arrives for T1
+ 1. The compaction of F4,F5,F6,F7 on EE2 is canceled and new a compaction of F4,F5,F6,F7,F8 is queued on EE2 for Tablet T1
+
+Scenario 2
+
+ 1. Tablet T1 has files F1,F2,F3,F4
+ 1. Compaction of files F1,F2,F3 is queued on internal executor IE1 for Tablet T1
+ 1. New files F5,F6,F7,F8,F9 arrive for tablet T1
+ 1. The compaction of F1,F2,F3 on IE1 is canceled and a new compaction of F1,F2,F3,F4,F5,F6,F7,F8,F9 is queued on external Executor EE2 for Tablet T1
+
+Scenario 3
+1. Tablet T1 has files F1,F2,F3
+2. Compaction of files F1,F1,F3 is queued on external executor EE2 for T1
+3. The table containing tablet T1 is deleted
+4. A compactor tries to compact F1,F2,F3 for tablet T1 and hopefully fails cleanly
```

Comment:
I think it would be better to call cancel on any compactions if a table is deleted.